diff --git a/libprocessgroup/include/processgroup/sched_policy.h b/libprocessgroup/include/processgroup/sched_policy.h
index a18847e..9d16e3a 100644
--- a/libprocessgroup/include/processgroup/sched_policy.h
+++ b/libprocessgroup/include/processgroup/sched_policy.h
@@ -48,6 +48,7 @@ typedef enum {
     SP_TOP_APP = 5,
     SP_RT_APP = 6,
     SP_RESTRICTED = 7,
+    SP_HSAE_APP = 8,
     SP_CNT,
     SP_MAX = SP_CNT - 1,
     SP_SYSTEM_DEFAULT = SP_FOREGROUND,
diff --git a/libprocessgroup/profiles/task_profiles.json b/libprocessgroup/profiles/task_profiles.json
index 45d3c7c..4fb6efa 100644
--- a/libprocessgroup/profiles/task_profiles.json
+++ b/libprocessgroup/profiles/task_profiles.json
@@ -15,6 +15,11 @@
       "Controller": "cpuset",
       "File": "top-app/cpus"
     },
+    {
+      "Name": "HsaeAppCapacityCPUs",
+      "Controller": "cpuset",
+      "File": "hsae-app/cpus"
+    },
     {
       "Name": "MemLimit",
       "Controller": "memory",
@@ -394,7 +399,19 @@
         }
       ]
     },
-
+    {
+      "Name": "ServiceCapacityHsaeApp",
+      "Actions": [
+        {
+          "Name": "JoinCgroup",
+          "Params":
+          {
+            "Controller": "cpuset",
+            "Path": "hsae-app"
+          }
+        }
+      ]
+    },
     {
       "Name": "CameraServiceCapacity",
       "Actions": [
@@ -616,6 +633,10 @@
       "Name": "SCHED_SP_TOP_APP",
       "Profiles": [ "MaxPerformance", "MaxIoPriority", "TimerSlackNormal" ]
     },
+    {
+      "Name": "SCHED_SP_HSAE_APP",
+      "Profiles": [ "ServiceCapacityHsaeApp" ]
+    },
     {
       "Name": "SCHED_SP_SYSTEM",
       "Profiles": [ "ServicePerformance", "LowIoPriority", "TimerSlackNormal" ]
@@ -648,6 +669,10 @@
       "Name": "CPUSET_SP_RESTRICTED",
       "Profiles": [ "ServiceCapacityRestricted", "TimerSlackNormal" ]
     },
+    {
+      "Name": "CPUSET_SP_HSAE_APP",
+      "Profiles": [ "ServiceCapacityHsaeApp" ]
+    },
     {
       "Name": "Dex2OatBootComplete",
       "Profiles": [ "Dex2oatPerformance", "LowIoPriority", "TimerSlackHigh" ]
diff --git a/libprocessgroup/sched_policy.cpp b/libprocessgroup/sched_policy.cpp
index 169b1d3..ae15f2e 100644
--- a/libprocessgroup/sched_policy.cpp
+++ b/libprocessgroup/sched_policy.cpp
@@ -57,6 +57,8 @@ int set_cpuset_policy(int tid, SchedPolicy policy) {
             return SetTaskProfiles(tid, {"CPUSET_SP_SYSTEM"}, true) ? 0 : -1;
         case SP_RESTRICTED:
             return SetTaskProfiles(tid, {"CPUSET_SP_RESTRICTED"}, true) ? 0 : -1;
+	case SP_HSAE_APP:
+            return SetTaskProfiles(tid, {"CPUSET_SP_HSAE_APP"}, true) ? 0 : -1;
         default:
             break;
     }
@@ -109,6 +111,9 @@ int set_sched_policy(int tid, SchedPolicy policy) {
         case SP_RT_APP:
             SLOGD("RT  tid %d (%s)", tid, thread_name);
             break;
+	case SP_HSAE_APP:
+            SLOGD("hsae app xxxxx  tid %d (%s)", tid, thread_name);
+            break;
         default:
             SLOGD("??? tid %d (%s)", tid, thread_name);
             break;
@@ -128,7 +133,9 @@ int set_sched_policy(int tid, SchedPolicy policy) {
             return SetTaskProfiles(tid, {"SCHED_SP_SYSTEM"}, true) ? 0 : -1;
         case SP_RT_APP:
             return SetTaskProfiles(tid, {"SCHED_SP_RT_APP"}, true) ? 0 : -1;
-        default:
+        case SP_HSAE_APP:
+            return SetTaskProfiles(tid, {"SCHED_SP_HSAE_APP"}, true) ? 0 : -1;
+	default:
             return SetTaskProfiles(tid, {"SCHED_SP_DEFAULT"}, true) ? 0 : -1;
     }
 
@@ -178,6 +185,8 @@ static int get_sched_policy_from_group(const std::string& group, SchedPolicy* po
         *policy = SP_TOP_APP;
     } else if (group == "restricted") {
         *policy = SP_RESTRICTED;
+    } else if (group == "hsae-app") {
+        *policy = SP_HSAE_APP;
     } else {
         errno = ERANGE;
         return -1;
@@ -234,7 +243,7 @@ const char* get_sched_policy_name(SchedPolicy policy) {
     static const char* const kSchedPolicyNames[] = {
             [SP_BACKGROUND] = "bg", [SP_FOREGROUND] = "fg", [SP_SYSTEM] = "  ",
             [SP_AUDIO_APP] = "aa",  [SP_AUDIO_SYS] = "as",  [SP_TOP_APP] = "ta",
-            [SP_RT_APP] = "rt",     [SP_RESTRICTED] = "rs",
+            [SP_RT_APP] = "rt",     [SP_RESTRICTED] = "rs", [SP_HSAE_APP] = "ct",
     };
     static_assert(arraysize(kSchedPolicyNames) == SP_CNT, "missing name");
     if (policy < SP_BACKGROUND || policy >= SP_CNT) {
@@ -255,7 +264,9 @@ const char* get_cpuset_policy_profile_name(SchedPolicy policy) {
     static constexpr const char* kCpusetProfiles[SP_CNT + 1] = {
             "CPUSET_SP_DEFAULT", "CPUSET_SP_BACKGROUND", "CPUSET_SP_FOREGROUND",
             "CPUSET_SP_SYSTEM",  "CPUSET_SP_FOREGROUND", "CPUSET_SP_FOREGROUND",
-            "CPUSET_SP_TOP_APP", "CPUSET_SP_DEFAULT",    "CPUSET_SP_RESTRICTED"};
+            "CPUSET_SP_TOP_APP", "CPUSET_SP_DEFAULT",    "CPUSET_SP_RESTRICTED",
+            "CPUSET_SP_HSAE_APP"
+    };
     if (policy < SP_DEFAULT || policy >= SP_CNT) {
         return nullptr;
     }
@@ -274,7 +285,7 @@ const char* get_sched_policy_profile_name(SchedPolicy policy) {
     static constexpr const char* kSchedProfiles[SP_CNT + 1] = {
             "SCHED_SP_DEFAULT", "SCHED_SP_BACKGROUND", "SCHED_SP_FOREGROUND",
             "SCHED_SP_SYSTEM",  "SCHED_SP_FOREGROUND", "SCHED_SP_FOREGROUND",
-            "SCHED_SP_TOP_APP", "SCHED_SP_RT_APP",     "SCHED_SP_DEFAULT"};
+            "SCHED_SP_TOP_APP", "SCHED_SP_RT_APP",     "SCHED_SP_DEFAULT", "SCHED_SP_HSAE_APP"};
     if (policy < SP_DEFAULT || policy >= SP_CNT) {
         return nullptr;
     }
diff --git a/rootdir/init.rc b/rootdir/init.rc
index f9db37d..052ab4f 100644
--- a/rootdir/init.rc
+++ b/rootdir/init.rc
@@ -448,6 +448,14 @@ on init
     chmod 0664 /dev/cpuset/camera-daemon/cgroup.procs
     chmod 0664 /dev/cpuset/quick-task/cgroup.procs
 
+    mkdir /dev/cpuset/hsae-app
+    copy /dev/cpuset/cpus /dev/cpuset/hsae-app/cpus
+    copy /dev/cpuset/mems /dev/cpuset/hsae-app/mems
+    chown system system /dev/cpuset/hsae-app/cgroup.procs
+    chmod 0664 /dev/cpuset/hsae-app/tasks
+    chmod 0664 /dev/cpuset/hsae-app/cgroup.procs
+    write /dev/cpuset/hsae-app/cpus 4-5
+
     # make the PSI monitor accessible to others
     chown system system /proc/pressure/memory
     chmod 0664 /proc/pressure/memory
